### ISSUE TITLE
type-value relationship not checked in generic function with multiple type args

### ISSUE THREAD

**Issue by Pwuts** [microsoft/pyright#10293]
Consider this function and a possible usage:
```py
def typed_cast(to_type: type[TT], from_type: type[T], value: T) -> TT:
    return cast(TT, value)

n: bool = typed_cast(bool, str, 123)  # no error here :(
```

I would expect Pyright to report an error when the type of `value` doesn't match `from_type`, but it doesn't.

I apologize in advance if this is already covered by an existing issue, but I had no idea what to search for. :')

**Update:** this *does* work:
```py
def typed(type: type[T], value: T) -> T:
    """
    Add an explicit type to a value. Useful in nested statements, e.g. dict literals.
    """
    return value


a: str = typed(str, 123)
# Expression of type "str | int" is incompatible with declared type "str"
#   Type "str | int" is incompatible with type "str"
#     "int" is incompatible with "str"
```
So it seems to be an issue specifically with functions with multiple type arguments.

**Comment by erictraut** [comment]
Pyright is working correctly here and in conformance with the typing spec, so this isn't a bug. There are no constraints between distinct type variables. They are solved independently of each other. There is nothing that requires the type for `TT` to be the same as the type for `T`. Not surprisingly, mypy and other type checkers likewise emit no error here.

The fact that you are using a `cast` here to override an error emitted by the type checker should be a good clue that you're doing something incorrect here.

**Comment by Pwuts** [comment]
> There are no constraints between distinct type variables. They are solved independently of each other. There is nothing that requires the type for `TT` to be the same as the type for `T`.

I'm not sure you read the example correctly. The issue is that `value: T` isn't checked against `from_type: type[T]`: two parameters both using **the same type variable `T`**. Of course it makes sense that `T` and `TT` are checked independently of each other, and I should have chosen a different letter to avoid this confusion.

**Comment by erictraut** [comment]
You're making an incorrect assumption about how constraint solving works. When a type checker solves a type variable, it collects a set of constraints and then attempts to find a type that satisfies all of these constraints. If no such type exists, then it issues a type error.

Here's a modified sample that is self-contained and minimal (doesn't contain additional superfluous aspects).

```python
def func[T](a: type[T], b: T) -> T: ...

n = func(str, 123)
reveal_type(n) # str | int
```

The function call in this example introduces two constraints: `T :> str` and `T :> int`. There is no type error here because the constraint solver is able to find a type that satisfies both of these constraints: `T = str | int`.

You'll see that mypy likewise accepts this code without error, although it picks the solution of `object` rather than `str | int`. Both solutions are technically correct, but pyright's is [more precise](https://microsoft.github.io/pyright/#/mypy-comparison?id=unions-vs-joins).

If one of these type variables were in an invariant position, the result would be different. Consider this modified code sample that uses `list[T]` in the annotation for the second parameter.

```python
def func[T](a: type[T], b: list[T]) -> T: ...

n = func(str, [123]) # Error
```

This is because the constraints are now `T :> str` and `T := int`. There is no type T that satisfies both of these constraints.

**Comment by Pwuts** [comment]
This is very helpful context. Thanks for taking the time to educate me.

Is there a way to achieve the behavior I expected? I guess not since (afaik) there is no way to say which occurrence of `T` should be used to infer `T` and which occurrences shouldn't.

**Comment by erictraut** [comment]
> Is there a way to achieve the behavior I expected?

Not if all uses of the type variable are in covariant (or contravariant) contexts. If one of them is in an invariant location, like in my example above, that adds more rigid constraints.
